[10.0] l10n_it_fatturapa: xml doctor for fatturapa

sherpya opened the pull request:
proposal per #1172 

it fixes:
 - removes DataOraConsegna, DataOraRitiro if dates are bogus
   i.e. 0001-01-01T00:00:00.000+02:00 where python raises
   OverflowError
 - removes timezone from Date to make pyxb able to compare with
   1-1-1970, it also removes the need of patching pyxb

breaking change:
  modules needs to import binding.fatturapa instead of
  bindings.fatturapa_v_1_2, this would be usefull also for
  new specs

--
Confermo di aver firmato il CLA https://odoo-community.org/page/cla e di aver letto le linee guida su https://odoo-community.org/page/contributing

sherpya commented:
@eLBati travis ha ricominciato a rompersi su roba che non ci incastra

sherpya commented on l10n_it_fatturapa/bindings/fatturapa.py:
```diff
@@ -0,0 +1,36 @@
+        root = etree.fromstring(xml_string)
+    except Exception as e:
+        _logger.warn('lxml was unable to parse xml: %s' % e)
+        return _CreateFromDocument(xml_string)
```
qui non è proprio bellissimo ma bisognerebbe cambiare tutti i test

eLBati commented on l10n_it_fatturapa/bindings/fatturapa.py:
```diff
@@ -0,0 +1,36 @@
+
+def CreateFromDocument(xml_string):
+    try:
+        root = etree.fromstring(xml_string)
```
Questo non solleva già OverflowError?

sherpya replied:
no parsa solo l'xml non guarda l'xsd

eLBati commented on l10n_it_fatturapa/bindings/fatturapa.py:
```diff
@@ -0,0 +1,36 @@
+        return _CreateFromDocument(xml_string)
+
+    # remove bogus dates accepted by ADE but not by python
+    for element in root.xpath('//DataOraConsegna | //DataOraRitiro'):
```
Hai già verificato che non ci siano altri elementi che contengono l'ora?

sherpya replied:
sono i soli xsd:dateTime, avrei preferito estrarre tutti i campi di quel tipo da pyxb ma non sono riuscito a capire come fare

eLBati replied:
Anche `Data` sembra avere il problema: https://github.com/OCA/l10n-italy/issues/1228

eLBati commented:
> travis ha ricominciato a rompersi su roba che non ci incastra

se i problemi non riguardano righe modificate da te, puoi ignorarli

sherpya commented:
> > travis ha ricominciato a rompersi su roba che non ci incastra
> 
> se i problemi non riguardano righe modificate da te, puoi ignorarli

è abbastanza randomico

sherpya commented:
@eLBati mi sembra ok, ora sistemo il test per controllare il risultato
```
2019-05-07 21:51:35,903 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: removed timezone information from date only 2014-12-18
2019-05-07 21:51:35,904 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: removed timezone information from date only 2015-07-01
2019-05-07 21:51:35,905 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: removed invalid dateTime 0001-01-01T00:00:00.000+02:00 for DataOraConsegna: date value out of range
2019-05-07 21:51:35,905 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: mandatory element Descrizione was empty, replacing with a dash
2019-05-07 21:51:35,908 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: removed empty Causale element
2019-05-07 21:51:35,909 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: removed empty NumeroCivico element
2019-05-07 21:51:36,081 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: removed timezone information from date only 2014-12-18
2019-05-07 21:51:36,082 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: removed timezone information from date only 2015-07-01
2019-05-07 21:51:36,083 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: removed invalid dateTime 0001-01-01T00:00:00.000+02:00 for DataOraConsegna: date value out of range
2019-05-07 21:51:36,083 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: mandatory element Descrizione was empty, replacing with a dash
2019-05-07 21:51:36,086 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: removed empty Causale element
2019-05-07 21:51:36,087 6969 WARNING openerp_test odoo.addons.l10n_it_fatturapa.bindings.fatturapa: removed empty NumeroCivico element
```

eLBati commented on l10n_it_fatturapa/bindings/xsd/FatturaPA_versione_1.2.1.xsd.diff:
```diff
@@ -0,0 +1,54 @@
+--- Schema_del_file_xml_FatturaPA_versione_1.2.1.xsd.orig	2019-02-12 12:10:42.024629978 +0100
```
Il diff serve ancora?

sherpya replied:
si perché purtroppo ci sono ancora cose fuori specifica e non si posso risolvere da xml,
tipo i numeri coi trailing zero

sherpya commented on l10n_it_fatturapa/bindings/fatturapa.py:
```diff
@@ -0,0 +1,135 @@
+    try:
+        root = etree.fromstring(xml_string)
+    except Exception as e:
+        _logger.warn('lxml was unable to parse xml: %s' % e)
```
al momento rimpalla a pyxb quindi ci sarà l'errore suo, sarebbe meglio dare errore qui ma i test si aspettano errori differenti, comunque se vuoi lo alzo a error

eLBati replied:
Ok per lasciare così

sherpya commented on l10n_it_fatturapa/bindings/fatturapa.py:
```diff
@@ -0,0 +1,135 @@
+    collect_elements_by_type_query(datetime_types, "//*[@type='xs:dateTime']")
+
+
+_xsd_schema = os.path.join(os.path.dirname(__file__), 'xsd', XSD_SCHEMA)
```
non sapevo, provvederò

sherpya replied:
però stavo pensando a parte questa fatturapa non ha dipendenze da odoo, potrebbe essere utile

eLBati replied:
Tipo per?

sherpya replied:
boh per fare test singoli, comunque nessun problema

sherpya commented:
> Anche
> 
> ```
> l10n_it_fatturapa/bindings/fatturapa.py:8:1: F403 'from binding import *' used; unable to detect undefined names
> l10n_it_fatturapa/bindings/fatturapa.py:42:20: F405 '_root' may be undefined, or defined from star imports: binding
> l10n_it_fatturapa/bindings/fatturapa.py:44:23: F405 '_root' may be undefined, or defined from star imports: binding
> l10n_it_fatturapa/bindings/fatturapa.py:49:20: F405 '_root' may be undefined, or defined from star imports: binding
> l10n_it_fatturapa/bindings/fatturapa.py:51:23: F405 '_root' may be undefined, or defined from star imports: binding
> l10n_it_fatturapa/bindings/fatturapa.py:57:25: F405 '_root' may be undefined, or defined from star imports: binding
> l10n_it_fatturapa/bindings/fatturapa.py:58:66: E261 at least two spaces before inline comment
> l10n_it_fatturapa/bindings/fatturapa.py:80:23: F405 'CreateFromDocument' may be undefined, or defined from star imports: binding
> ```

non posso mettere a mano 500 imports :D c'è posso dire a flake di ignorare la cosa? per E261 è una # di refuso

eLBati commented:
> non posso mettere a mano 500 imports

Come mai 500? Non basterebbero `_root` e `CreateFromDocument`?

sherpya commented:
> > non posso mettere a mano 500 imports
> 
> Come mai 500? Non basterebbero `_root` e `CreateFromDocument`?

parlavo del wild import (ho messo ignore), _root è stupido flake non si accorge che lo assegno sotto

sherpya commented:
si è intrippato travis, da errori differenti.. :(

eLBati commented:
Che in locale non ottieni?

sherpya commented:
> Che in locale non ottieni?

In locale ho sempre casino, cmq ho ri-pushato credo che aggiungesse un '\n\n' alle inconsistenze che a qualche test non piaceva

sherpya commented:
sembra ok ora, almeno la mia roba

eLBati commented on l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py:
```diff
@@ -1209,6 +1209,10 @@ def importFatturaPA(self):
 
+            xmlproblems = getattr(fatt, '_xmldoctor', None)
+            if xmlproblems:  # None or []
+                generic_inconsistencies += '\n'.join(xmlproblems) + '\n\n'
```
Potresti solo aggiungere un pezzo di test che controlli che nella fattura odoo ci sia ad esempio "removed timezone..."?

sherpya replied:
ok aggiunto

SimoRubi commented on l10n_it_fatturapa/bindings/fatturapa.py:
```diff
@@ -0,0 +1,147 @@
+
+    path = '//%s/%s' % (parent.attrib['name'], element.attrib['name'])
+    mandatory = element.attrib.get('minOccurs') != '0'
+    assert path not in target or target[path] == mandatory
```
Non mi è ben chiaro cosa dovrebbe fare questo `assert`, forse sarebbe più trasparente usando degli `if`?

sherpya replied:
è un check di debug, si può anche levare

sherpya replied:
in realtà serve per vedere se non lo sto sostituendo con lo stesso elemento ma differente minOccurs, lo metto in modo diverso

SimoRubi replied:
Ah ok pensavo fosse una specie di filtro, non sapevo cosa facesse `assert`.

Un altro dubbio che avevo su questa parte è: qui `element` è un elemento ben preciso nel XML, quindi perché non prendi il suo path con `getpath`?
Dapprima pensavo che fosse perché poi più tardi trovi tutti i nodi che hanno il path `//parent_name/element_name`, però riguardandolo adesso, non è detto che tutti quegli elementi siano `mandatory`.
Mi sfugge qualcosa?

sherpya replied:
in verità ho preso solo il parent, non credo sia necessario prendere tutto

sherpya replied:
getpath non so se funziona qui, perché questo è l'xsd io ho da trovare il tag nella fattura invece

SimoRubi replied:
Quindi comunque ci aspettiamo che venga trovato solo un elemento da `//parent_name/element_name`?

SimoRubi replied:
oppure tutti quelli trovati saranno comunque `mandatory`?

sherpya replied:
> Quindi comunque ci aspettiamo che venga trovato solo un elemento da `//parent_name/element_name`?

o se ce ne sono più di uno che abbia lo stesso valore per mandatory, cioè se lo sovrascrivo ed è obbligatorio o non obbligatorio in entrambi mi va bene uguale

SimoRubi replied:
Ma se ce ne fossero due con valori diversi non genererebbe problemi sovrascriverli?
Se ad esempio da `root.xpath('//parent_name/element_name')` risultassero due elementi:

- **e1** obbligatorio
- **e2** facoltativo

Sovrascrivendo, può risultare `target['//parent_name/element_name'] = False`.
Questo rimuoverebbe quindi **e1** che non va rimosso.

Probabilmente nei file XML che stai gestendo questo non succede, ma considerato che in sti file sta arrivando di tutto, meglio premunirsi

SimoRubi replied:
Guardando l'`xsd` questo non succede, e capendo un po' meglio come funzionano, direi che è difficile che succeda quindi direi che va bene anche così. Se ci saranno problemi vedremo che fare

sherpya replied:
Non succede con questo xsd, se domani arriva un 1.3 potrebbe succedere, l'assert è lì per quello, e ci si accorgerebbe subito

SimoRubi commented on l10n_it_fatturapa/bindings/fatturapa.py:
```diff
@@ -0,0 +1,147 @@
+        target[path] = mandatory
+
+
+def collect_elements_by_type(target, element_type):
```
Questo metodo potrebbe essere ridotto a una chiamata del successivo:
```
def collect_elements_by_type(target, element_type):
    collect_elements_by_type_query(target, get_type_query(element_type)):
```

sherpya replied:
si è vero, a fare refactor non mi sono accorto :D

SimoRubi commented on l10n_it_fatturapa/bindings/fatturapa.py:
```diff
@@ -0,0 +1,147 @@
+                result = result.replace(tzinfo=None)
+                element.text = result.XsdLiteral(result)
+                msg = 'removed timezone information from date only element ' \
+                      '%s: %s' % (path[2:], element.text)
```
Questo `path` è comune anche ad altri elementi, si potrebbe usare https://lxml.de/api/lxml.etree._ElementTree-class.html#getpath per avere il suo percorso univocamente.

Stessa cosa per le altre occorrenze sotto

sherpya replied:
fatto però esce una roba così:
`/p:FatturaElettronica/FatturaElettronicaBody/DatiGenerali/DatiTrasporto/IndirizzoResa/NumeroCivico element`, ci può andare bene uguale credo

SimoRubi commented on l10n_it_fatturapa/bindings/fatturapa.py:
```diff
@@ -0,0 +1,147 @@
+            if mandatory:
+                msg = 'mandatory element %s was empty, replacing with a dash' \
+                      % path[2:]
+                problems.append(msg)
```
Potrebbe essere spostato dopo l'`if/else`, per evitare anche la ripetizione sotto

SimoRubi commented on l10n_it_fatturapa/bindings/fatturapa.py:
```diff
@@ -0,0 +1,149 @@
+        target[path] = mandatory
+    else:
+        assert target[path] == mandatory, 'Element already present with ' \
+                                          'different minOccurs valus'
```
```suggestion
                                          'different minOccurs value'
```
Si potrebbe anche aggiungere il `path` (e volendo anche il `text`) per uniformarlo agli altri messaggi di warning e poter capire dai log che sta succedendo

sherpya replied:
ma è un assert, cmq dimmi cosa scrivo, ce lo metto, non dovrebbe capitare

SimoRubi replied:
```
        if target[path] != mandatory:
            msg = "Element having path %s already present with a different minOccurs value" % path
            _logger.warn(msg)
```

sherpya replied:
no deve proprio scoppiare se capita :)

SimoRubi replied:
Va bene, secondo me è utile indicare il path di cosa sta fallendo nei log per capire cosa sta succedendo e dove.
Se usarlo per sollevare eccezione o con un assert decidi tu

sherpya replied:
si ora riproduco aggiungendo un doppione nell'xsd e vedo come viene meglio l'errore

sherpya replied:
credo una cosa simile sia accettabile:
`Element //DatiTrasporto/Descrizione is already present with different minOccurs value`

eLBati commented:
@SimoRubi @sherpya quindi procedo con il merge?

sherpya commented:
> @SimoRubi @sherpya quindi procedo con il merge?

le modifiche richieste sono state effettuate, se non ne avete altre puoi procedere
